src/board.rs

```diff
@@ -567,6 +562,40 @@ impl Board {
         self.move_history.push(tuple.clone());
     }
 
+    fn hist_to_coord(hist_item: &str) -> Vec<usize> {
+        assert_eq!(hist_item.chars().count(), 2);
+        vec![
+            hist_item
+                .chars()
+                .next()
+                .unwrap()
+                .to_string()
+                .parse()
+                .unwrap(),
+            hist_item
+                .chars()
+                .nth(1)
+                .unwrap()
+                .to_string()
+                .parse()
+                .unwrap(),
+        ]
+    }
+
+    pub fn undo(&mut self) {
+        if let Some(prev_hist) = self.move_history.pop() {
+            let prev_move = prev_hist.1;
+
+            let to = Self::hist_to_coord(&prev_move[0..2]);
+            let from = Self::hist_to_coord(&prev_move[2..]);
+            self.board[to[0]][to[1]] = self.board[from[0]][from[1]];
+            // self.board[to[0]][to[1]] = self.board[from[0]][from[1]];
+            // self.board[to[0]][to[1]] = Some((PieceType::Pawn, PieceColor::Black));
+            self.board[from[0]][from[1]] = None;
+            self.switch_player_turn();
+        }
+    }
```

Comment:
This is a great idea to implement, but it appears to be missing some cases. For instance, when a piece is captured, it needs to be placed back on the board. Additionally, when the bot makes a move, we might not want to allow undoing, or we should thoroughly review the undo functionality as it seems to be behaving unexpectedly at the moment !

Reply:
Yes, I totally agree. In fact I've already written placing taken off pieces back and will commit it right now. It's not yet tested, but seems to be working quite well. I don't really have an idea about how to deal with the Bot.

Reply:
Maybe we can just desactivate the undo if it's playing with against the bot ?